MMA-izing the prolongator and restrictor kernels

hummingtree opened the pull request:
As the name suggests, this PR adds initial support for MMA-izing the prolongator and restrictor kernels. In addition,
- Initial supports are also added to use tensor memory accelerator (TMA) for the memory movement on the Hopper and Blackwell GPUs;
- Some general shared memory load/store patterns are optimized.
- As a general cleanup, the MMA types can now be specified as a CMake parameter, e.g.,
```
cmake ...
  -DQUDA_MULTIGRID_MMA_DSLASH_TYPE_HALF=2 \
  -DQUDA_MULTIGRID_MMA_PROLONGATOR_TYPE_HALF=3 \
  -DQUDA_MULTIGRID_MMA_RESTRICTOR_TYPE_HALF=3 \
  -DQUDA_MULTIGRID_MMA_SETUP_TYPE_HALF=0 \
  -DQUDA_MULTIGRID_MMA_DSLASH_TYPE_SINGLE=2 \
  -DQUDA_MULTIGRID_MMA_PROLONGATOR_TYPE_SINGLE=3 \
  -DQUDA_MULTIGRID_MMA_RESTRICTOR_TYPE_SINGLE=3 \
  -DQUDA_MULTIGRID_MMA_SETUP_TYPE_SINGLE=0 \
  ...
```
The encoding is the following: 
```
    "1->SIMT; 2->SMMA; 3->1xFP16; 4->3xFP16; 5->1xTF32; 6->3xTF32; 7->3xBF16; 0->DEFAULT
```
The default types are:
|    | half/default | single/default
| :-: | :-: | :-: |
| Setup  | 3xFP16    | 3xFP16 |
| Coarse Dslash | 3xBF16 (SIMT for < SM80) | 3xTF32 (SIMT for < SM80) |
| Restrictor   | 3xBF16 (SIMT for < SM80) | 3xTF32 (SIMT for < SM80) |
| Prolongator | 3xBF16 (SIMT for < SM80) | 3xTF32 (SIMT for < SM80) |

- For coarse dslash, prolongator and restrictor, the code will automatically find the suitable `nVec` instantiation to use, e.g., if `nVec = 16,32` are instantiated, for `nRHS = 5`, `nVec = 16` will be picked; for `nRHS = 24`, `nVec = 32` will be picked; for `nRHS = 96`, the `nVec = 32` kernel will be called 3 times to divide and conquer.

Remaining to-dos:
- [x] Add a command line interface for toggling MMA-ized prolongator and restrictor kernels.
- [x] Make the prolongator and restrictor working for staggered (the spin/parity handling is different).
- [x] A comprehensive testing of the MG/MMA workflows.
- [x] Add doxygen.
- [ ] Clang-format

![coarse_dslash_tma](https://github.com/user-attachments/assets/75165096-601c-40e9-9605-1ae6554042e9)

![transfer_mma](https://github.com/user-attachments/assets/2dacaec7-9555-4eea-9028-a6c1a6565544)

hummingtree committed:
Add MMA version of prolongator.

hummingtree committed:
Dagger'ed the equation to make better use of MMA.

hummingtree committed:
Make nColor = 3 works:
- Still need to add the spin factor of 2;
- Still need to cover the to_non_rel = true;

hummingtree committed:
Add from_to_non_rel to block transpose to complete the circle.

hummingtree committed:
More cleanup of the MMA code. Apply vector gmem loads when possible.

hummingtree committed:
Apply more vector gmem loads when possible.

hummingtree committed:
Add MMA version for restrictor.

hummingtree committed:
Add expands for restrictor with MMA.

hummingtree committed:
Add shared memory caching for the the restrictor kernel.

hummingtree committed:
Allow restrictor_mma to have N % bN != 0; More generic optimizations for loading from gmem.

hummingtree committed:
Add rescaling to prolongator.

hummingtree committed:
Modify the MMA types.

hummingtree committed:
Add rescale for restrictor.

hummingtree committed:
Abstract the MMA expansions into a class.

hummingtree committed:
Add more abstraction; make TF32 the default for SM80 and later.

maddyscientist committed:
Merge branch 'feature/mrhs-solvers' of github.com:lattice/quda into feature/prolongator-mma

hummingtree committed:
Merge branch 'feature/mrhs-solvers' of github.com:lattice/quda into feature/prolongator-mma

hummingtree committed:
Fix block transpose by having no bound checks on the kernel level; in the process add an additional template to `kernel_param`.

maddyscientist committed:
Merge branch 'feature/prolongator-mma' of github.com:lattice/quda into feature/prolongator-mma

maddyscientist committed:
Add some aggregate sizes to MMA restrictor

hummingtree committed:
Make `aggregate_size` a runtime variable.

hummingtree committed:
Apply clang-format.

hummingtree committed:
Soften the restriction for nrhs from multiple of 16 to multiple of 8.

hummingtree committed:
Set the default precision in coarse dslash mma to TF32/FP16; Fix the underlying code such that FP16 works with rescaling.

hummingtree committed:
Clean up the MMA code.

hummingtree committed:
Short cut the rescaling code to use scale_inv for fixed point format fields.

hummingtree committed:
Clean up code.

hummingtree committed:
Merge branch 'feature/mrhs-solvers' of github.com:lattice/quda into feature/prolongator-mma

maddyscientist commented:
Thanks @hummingtree.  Noting this PR is dependent on #1489.

hummingtree committed:
Merge branch 'feature/mrhs-solvers' of github.com:lattice/quda into feature/prolongator-mma

hummingtree committed:
Add const and constexpr; add nrhs to prolongator tuning string.

hummingtree committed:
Add initial support for TMA in coarse dslash; optimize the shared memory patterns accordingly.

hummingtree committed:
Add more shared memory patterns.

hummingtree committed:
Apply TMA to the clover term; clean up code and remove duplications.

hummingtree committed:
Relax the requirements in multigrid_benchmark_test.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/prolongator-mma

hummingtree committed:
Put tma_wait under the TMA macro.

hummingtree committed:
Add const to `include/kernels/dslash_mdw_fused.cuh`; Put `lib/prolongator_mma.in.cu` and `lib/restrictor.in.cpp` under the `QUDA_MMA_AVAILABLE` macro.

hummingtree committed:
Add more const to Arg &arg; Fix compile when MMA is not available.

hummingtree committed:
Fix CI:
- Fix the issue with modifying the kenrel argument in fused MDW kernels;
- Fix the issue where QUDA_ENABLE_MMA not being an option when MULTIGRID is disabled;
- Fix compilation when MMA is not enabled.

hummingtree committed:
Remove the unwanted warning macro.

hummingtree committed:
Address the unused parameter warnings from CI.

hummingtree committed:
Add support for staggered nSpin = 1 case for restrictor_mma.

hummingtree committed:
Break TMA box sizes into smaller pieces when the sizes are larger than 256.

hummingtree committed:
Restore the SMMA precision for coarse dslash MMA.

hummingtree committed:
Use SIMT for coarse dslash MMA on SM70.

hummingtree committed:
Add param and command line option for running transfer with MMA.

hummingtree committed:
Add doxygen to `include/expand_list.hpp` and `include/targets/cuda/tma_helper.hpp`.

weinbe2 commented on include/color_spinor_field_order.h:
```diff
@@ -858,6 +858,8 @@ namespace quda
       static constexpr int nSpin = nSpin_;
       static constexpr int nColor = nColor_;
 
+      using store_type = storeFloat;
```
I believe this should be `store_t` to match the `real_t`, etc, convention we're slowly working towards---but if there's precedent for `store_type` elsewhere I won't fight too hard about this.

hummingtree replied:
I can convert all the `store_type` to `store_t` - they are only used by the MMA-related code - should I just do it?

hummingtree replied:
https://github.com/lattice/quda/commit/d0da82373faf185f7b11fee05b7415cdbecf8424

weinbe2 commented on include/targets/cuda/mma_tensor_op/smma_m16n8k8_sm70.cuh:
```diff
@@ -0,0 +1,150 @@
+
+      static constexpr bool do_rescale()
+      {
+        return true; // false because we use FP16
```
typo on the comment

hummingtree replied:
https://github.com/lattice/quda/commit/d0da82373faf185f7b11fee05b7415cdbecf8424

weinbe2 commented on include/targets/cuda/tma_helper.hpp:
```diff
@@ -0,0 +1,213 @@
+    if (CUDA_SUCCESS != error) {
+      const char *str;
+      cuGetErrorName(error, &str);
+      errorQuda("TMA descriptor creation returned %s\n", str);
```
No need for a `\n`

hummingtree replied:
https://github.com/lattice/quda/commit/d0da82373faf185f7b11fee05b7415cdbecf8424

weinbe2 commented on lib/block_transpose.in.cu:
```diff
@@ -104,82 +115,70 @@ namespace quda
       errorQuda("QUDA_PRECISION=%d does not enable required precision combination (V = %d B = %d)", QUDA_PRECISION,
                 V.Precision(), B[0].Precision());
 
+    IntList<1, 2, 4> nSpins;
```
I know there isn't a huge file to begin with, but this could be thinned down depending on which Dirac types are compiled in.

hummingtree replied:
You mean this file can be thinned down depending on nSpin? The code here is agnostic to Dirac types.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/cd3bbad7179f780979e7b6ffe2a88f71a5f18046

weinbe2 commented on lib/prolongator.in.cpp:
```diff
@@ -1,50 +1,118 @@
+
+          bool to_non_rel = (out.Nspin() == 4) && (out[0].GammaBasis() == QUDA_UKQCD_GAMMA_BASIS);
+          BlockTransposeBackward(v_out, out, to_non_rel);
+#if 0
```
Is there a reason to keep this around or can it be deleted?

hummingtree replied:
These are essentially the only test code for prolonator and restrictor (with/without MMA) - we might want to have a separate test/ctest for prolongator and restrictor.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/1be04f52b1f513c9e5a522a759884f736329e487

hummingtree committed:
Fix typo; replace `store_type` with `store_t`.

hummingtree committed:
Only instantiate enabled spins in `lib/block_transpose.in.cu`.

hummingtree committed:
Remove test code in `lib/restrictor.in.cpp` and `lib/prolongator.in.cpp`.

weinbe2 commented on include/expand_list.hpp:
```diff
@@ -0,0 +1,200 @@
+namespace quda {
+
+/**
+    @brief Thie helper class instantiates the following mapping:
```
`Thie` -> `This`

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/6451e7fb56974b902baacdb64f547ef8eb5231c4

weinbe2 commented on include/kernels/dslash_coarse_mma.cuh:
```diff
@@ -12,6 +12,14 @@
 #include <mma_tensor_op/gemm.cuh>
 #include <complex_quda.h>
 
+#if (__COMPUTE_CAPABILITY__ >= 900) && (CUDA_VERSION >= 12060)
```
There's a bit of a design question here--in principle this type of thing should go into a target backend, but it feels like it would be an aggressive re-engineering. I still think we should pursue it, though. Thoughts @maddyscientist ?

weinbe2 replied:
I guess `*_mma.cuh` is already aggressively CUDA-specific, so maybe it doesn't matter so much for now.

hummingtree replied:
This is only compiled for CUDA in the first place so yeah I don't think it matters at this point.

maddyscientist replied:
Agree with @hummingtree here, we can defer any abstraction cleanup for now here, given this is a CUDA target only file.

weinbe2 commented on include/targets/cuda/kernel.h:
```diff
@@ -89,14 +89,19 @@ namespace quda
-    if (j >= arg.threads.y) return;
 
-    while (i < arg.threads.x) {
+    if constexpr (Arg::check_bounds) {
```
Should this go into, for ex, the `hip` target for completeness/consistency even if it's unused for now?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/6451e7fb56974b902baacdb64f547ef8eb5231c4

hummingtree replied:
Actually even for the CUDA target this `check_bounds` variable is only used in `kernel2D_impl` (i.e., not used in `kernel3D_impl`, etc). Adding similar changes to all the other kernels seems too aggressive though.

hummingtree replied:
@maddyscientist you might have a better way of doing this? So currently `check_bounds`/`check_bounds_` are added to `kernel_param`, and it is picked up only here in `Kernel2D_impl`, in the CUDA target.

maddyscientist replied:
I agree with @weinbe2 here, can you add this check to the other kernels (e.g., `Kernel3D_impl`, etc.) and probably also add it to the HIP variants?  Nice addition though.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/b0b435516b2b5f3fb76545960d5f65cfaa69eca6

weinbe2 commented on include/targets/cuda/mma_tensor_op/gmem_loader.cuh:
```diff
@@ -0,0 +1,792 @@
+            }
+            __syncthreads();
+
+            block_rescale_factor = 65504.0f / block_max_all; // 65504 = the maximum FP16 number
```
I don't love `65504.0f` just being a "magic number" here, but I don't think there's a better place to put it---elsewhere in QUDA we use `std::numeric_limits`, but there isn't a `half_t` yet (it's not coming until C++23). Maybe we should create a custom `quda::numeric_limits` that calls `std::numeric_limits` where it's C++ legal and then has custom overloads for `half`? Still awkward because we use `short` as a magic way to refer to `half`, I guess... thoughts?

weinbe2 replied:
Maybe it could just be something in `gmem_loader.cuh`?

maddyscientist replied:
Does libcu++ have this already?  It often back ports things to earlier, and may already support `numeric_limits` for type `T = __half` or some such.

hummingtree replied:
I have added `numeric_limits<half>` in `gmem_loader.cuh`, for now, in https://github.com/lattice/quda/pull/1497/commits/6451e7fb56974b902baacdb64f547ef8eb5231c4.

weinbe2 commented on lib/prolongator_mma.in.cu:
```diff
@@ -0,0 +1,244 @@
+#if (__COMPUTE_CAPABILITY__ >= 800)
+    using mma_t = hmma::hmma_tfloat32_t<4, 1, 1>;
+#else
+    using mma_t = hmma::hmma_x_t<16, 8, 8, mma::half, mma::half2>;
```
Should this be float instead of half? I'm just not sure of the convention here

hummingtree replied:
It's either TF32 or FP16 here so I think it's right - on SM70 we don't have TF32 so we have to use FP16.

hummingtree committed:
Store the aux values as the shape values directly (instead of as the indices).

hummingtree committed:
Use mma::numeric_limits<mma::half>::max() in MMA code.

hummingtree committed:
Add initializer to *_factors.

hummingtree committed:
Use curly brace initializers.

hummingtree committed:
Initialize looping variable.

hummingtree committed:
Specify MMA types in CMake.

hummingtree committed:
The MMA kernels will now choose smallest instantiated nVec that is larger than the actual nVec when creating the MMA-ordered fields.

hummingtree committed:
Add missing file; add checks for MRHS.

hummingtree committed:
Add comments to the newly added MMA macros.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/prolongator-mma

maddyscientist commented:
When we zero pad when using less RHS than what we support with the tensor core kernels, e.g., using 4 RHS with native 8 RHS tensor core support, can we make sure that the flops are counted correctly, and we don't count imaginary flops?

hummingtree committed:
Add nVec_actual to the color_spinor_fields/params - now the MMA kenrels' flops is the 'real' ones.

hummingtree committed:
Put the IntList into a header file.

maddyscientist commented on lib/targets/cuda/target_cuda.cmake:
```diff
@@ -198,12 +198,44 @@ set(QUDA_MULTIGRID_MRHS_LIST ${QUDA_MULTIGRID_MRHS_DEFAULT_LIST} CACHE STRING "T
 
-  option(QUDA_MULTIGRID_SETUP_USE_SMMA "Enabling using SMMA (3xTF32/3xBF16/3xFP16) for multigrid setup" ON)
-  mark_as_advanced(QUDA_MULTIGRID_SETUP_USE_SMMA)
+  set(QUDA_MULTIGRID_MMA_TYPES "SIMT" "SMMA" "1xFP16" "3xFP16" "1xTF32" "3xTF32" "3xBF16" "DEFAULT")
```
Can you check at `CMake` configure time that a valid MMA type is being used?  At the moment it will let you select `TF32` types and try to build even if you are on sm_70.  Best to catch this kind of error at configure time.

hummingtree replied:
I can add the check yeah.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/df97e1a881e0922287b4d39755c1262f92c32283

weinbe2 commented:
As far as I can tell it doesn't look like the MMA prolongator/restrictor is getting picked up with staggered on Hopper, at least using the default `cmake` parameters:

```
cmake -DQUDA_DIRAC_DEFAULT_OFF=ON \
      -DQUDA_DIRAC_STAGGERED=ON \
      -DCMAKE_BUILD_TYPE=DEVEL \
      -DQUDA_BACKWARDS=ON \
      -DCMAKE_INSTALL_PREFIX=/scratch/local/install \
      -DQUDA_MULTIGRID=ON \
      -DQUDA_MULTIGRID_NVEC_LIST="24,64,96" \
      -DQUDA_MULTIGRID_MRHS_LIST="8,16,32" \
      -DQUDA_DOWNLOAD_USQCD=ON \
      -DQUDA_QIO=ON -DQUDA_QMP=ON \
      -DQUDA_PRECISION=14 \
      -DQUDA_RECONSTRUCT=4 \
      -DQUDA_GPU_ARCH=sm_90 \
      /scratch/local/quda
```

Commands:
```
export QUDA_RESOURCE_PATH=`pwd`/tunecache

mpirun -np 1 ./heatbath_test --dim 16 16 16 16 --save-gauge l16t16b7p0   --heatbath-beta 7.0 --heatbath-coldstart true --heatbath-num-steps 10 --heatbath-warmup-steps 1000

mpirun -np 1 ./staggered_invert_test \
  --prec double --prec-sloppy single --prec-null half --prec-precondition half \
  --mass 0.01 --recon 18 --recon-sloppy 18 --recon-precondition 18 \
  --dim 16 16 16 16 --gridsize 1 1 1 1 --load-gauge l16t16b7p0 \
  --dslash-type asqtad --compute-fat-long true --tadpole-coeff 0.905160183 --tol 1e-10 \
  --verbosity verbose --solve-type direct --solution-type mat --inv-type gcr \
  --inv-multigrid true --mg-levels 4 --mg-coarse-solve-type 0 direct --mg-staggered-coarsen-type kd-optimized \
  --mg-block-size 0 1 1 1 1 --mg-nvec 0 3 \
  --mg-block-size 1 4 4 4 4 --mg-nvec 1 64 --mg-nvec-batch 1 32 \
  --mg-block-size 2 2 2 2 2 --mg-nvec 2 96 --mg-nvec-batch 2 32 \
  --mg-setup-tol 1 1e-5 --mg-setup-tol 2 1e-5 --mg-setup-inv 1 cgnr --mg-setup-inv 2 cgnr \
  --nsrc 32 --nsrc-tile 8 --niter 25 \
  --mg-setup-use-mma 0 true --mg-setup-use-mma 1 true --mg-setup-use-mma 2 true --mg-setup-use-mma 3 true \
  --mg-dslash-use-mma 0 true --mg-dslash-use-mma 1 true --mg-dslash-use-mma 2 true --mg-dslash-use-mma 3 true \
  --mg-transfer-use-mma 0 true --mg-transfer-use-mma 1 true --mg-transfer-use-mma 2 true --mg-transfer-use-mma 3 true \
  --mg-smoother 0 ca-gcr --mg-smoother-solve-type 0 direct  --mg-nu-pre 0 0 --mg-nu-post 0 4 \
  --mg-smoother 1 ca-gcr --mg-smoother-solve-type 1 direct --mg-nu-pre 1 0 --mg-nu-post 1 4 \
  --mg-smoother 2 ca-gcr --mg-smoother-solve-type 2 direct-pc  --mg-nu-pre 2 0 --mg-nu-post 2 4 \
  --mg-coarse-solver 1 gcr --mg-coarse-solve-type 1 direct --mg-coarse-solver-tol 1 0.25 --mg-coarse-solver-maxiter 1 16 \
  --mg-coarse-solver 2 gcr --mg-coarse-solve-type 2 direct-pc --mg-coarse-solver-tol 2 0.25 --mg-coarse-solver-maxiter 2 16 \
  --mg-coarse-solver 3 ca-gcr --mg-coarse-solve-type 3 direct-pc --mg-coarse-solver-tol 3 0.25 --mg-coarse-solver-maxiter 3 16 \
  --mg-verbosity 0 verbose --mg-verbosity 1 verbose --mg-verbosity 2 verbose --mg-verbosity 3 verbose
```

I can confirm the dslash and setup flags are being picked up correctly (but I was a bit surprised to see only `1xfp16` was getting used for the half-precision `CalculateYhat`, it felt a little low...), so I think it's uniquely the prolongator/restrictor that's having plumbing issues, at least for the default.

maddyscientist committed:
Some improvements to handle nVec and nVec_actual: nVec should be stored in aux key to ensure any degeneracy is broken; since nVec_actual doesn't affect allocations, we can simply set it correctly when return a field temp in FieldTmp<ColorSpinorField>.  Use a FieldCache for the transfer operator temporary to reduce overhead

maddyscientist committed:
Fix bytes comutation for MRHS clover solo operator

maddyscientist commented:
I have fixed an issue where I wasn't seeing `nVec_actual` show up in the tune string.  This was caused by the `FieldCache` key being ignorant of this variable (and indeed `nVec` itself), and so we could fetch a field from the cache for the reorder temporary from `create_color_spinor_copy` that had a different `nVec_actual` value in it.

maddyscientist commented on include/quda_define.h.in:
```diff
@@ -154,6 +154,52 @@
+
+#ifdef QUDA_MULTIGRID
+
+#cmakedefine QUDA_MULTIGRID_MMA_SETUP_TYPE
```
@weinbe2 and I discussed this a bit: the ideal here would be if we could set the MMA type used independently for single and half precision.  E.g., so we could use 3xTF32 for single-precision fields, while using 1xFP16, say, for half-precision fields.  If this is a pain to do we can defer this ask to a subsequent PR.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/254a79b5c6a31998ce4957a36d21eb654cfaddc3

maddyscientist commented on include/expand_list.hpp:
```diff
@@ -0,0 +1,206 @@
+namespace quda {
+
+/**
+    @brief This helper class instantiates the following mapping:
```
Wondering if we could clean up this file by switching `TuneParam::aux` to use `array<int, 4>` which would give an indexable accessor, which would avoid the need for all the `x`, `y`, `z`, `w` specializations?

hummingtree replied:
I think that would work nicely but would require a lot of code changes.

maddyscientist commented on include/expand_list.hpp:
```diff
@@ -0,0 +1,206 @@
+        if constexpr (sizeof...(Ws) > 0) {
+          span_w<Bx, By, Bz>(tp, stream, std::index_sequence<Ws...>());
+        } else {
+          errorQuda("Invalid tp.aux.w");
```
Add the invalid value to the print statement here (and below equivalents too)

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on include/kernels/dslash_coarse_mma.cuh:
```diff
@@ -92,9 +110,57 @@ namespace quda
     }
   };
 
+#ifdef USE_TENSOR_MEMORY_ACCELERATOR
```
Should this not be in tma_helper.hpp?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/68ab1406d1ec1aa135f509d5425f023e439bc97d

maddyscientist commented on include/kernels/dslash_mdw_fused.cuh:
```diff
@@ -199,7 +199,7 @@ namespace quda {
     */
     template <class storage_type, bool dagger, bool halo, bool back, class Vector, class Arg>
-    __device__ inline void apply_wilson_5d(Vector &out, int coordinate[4], Arg &arg, int s)
+    __device__ inline void apply_wilson_5d(Vector &out, int coordinate[4], const Arg &arg, int s)
```
A lot of function signatures have been changed in this PR have been changed to `const Arg&` from `Arg&`: while this is good to do, curious if there was a specific functional reason for this?

maddyscientist replied:
Answering my own question: I think it's because of the switch to using `__grid_constant__` for `raw_kernel`.

hummingtree replied:
Exactly.

maddyscientist commented on include/kernels/restrictor_mma.cuh:
```diff
@@ -0,0 +1,329 @@
+    loop_over<contiguous_dim, contiguous_limit, elements_per_thread>(
+      gmem, x_coarse, coarse_spin, contiguous_dim_offset, aggregate_k_offset, coarse_to_fine, arg, write_to_smem);
+
+    return 1.0f / block_rescale_factor;
```
I think we can avoid this division if we precompute `mma::numeric_limits<mma::half>::max()` on the host prior to calling the kernel, making this a multiply operation.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/53c8e31e71603e0bb70114ec4af6488a8eb5f828

maddyscientist commented on include/targets/cuda/mma_tensor_op/hmma_m16n16k4_sm70.cuh:
```diff
@@ -221,15 +221,19 @@ namespace quda
                    f2i_round<store_t>(-op_c_imag.reg[i * 2 + 0] * scale)};
-              op(&ptr[(n_index + 0) * ldc + m_index], s);
+              if (!check_bounds || (m_index < M && (n_index + 0) < N)) { op(&ptr[(n_index + 0) * ldc + m_index], s); }
+              // op(&ptr[(n_index + 0) * ldc + m_index], s);
```
delete comment lines (and below)

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on include/targets/cuda/mma_tensor_op/hmma_m16n8k8_sm70.cuh:
```diff
@@ -0,0 +1,277 @@
+              s = {f2i_round<store_t>(op_c_real.reg[i * 2 + 0] * scale),
+                   f2i_round<store_t>(-op_c_imag.reg[i * 2 + 0] * scale)};
+              if (!check_bounds || (m_index < M && (n_index + 0) < N)) { op(&ptr[(n_index + 0) * ldc + m_index], s); }
+              // op(&ptr[(n_index + 0) * ldc + m_index], s);
```
delete comments lines (and below)

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on include/transfer.h:
```diff
@@ -295,17 +300,19 @@ namespace quda {
        - V: spatial -> spin/color -> nVec
      @param[out] The output V Matrix field
      @param[in] B input vectors
+     @param[in] from_non_rel whether or not transform B from non-reletivistic basis
```
relativistic (typo)

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on include/targets/cuda/tma_helper.hpp:
```diff
@@ -0,0 +1,213 @@
+  {
+    CUtensorMap ret_value;
+
+    cuuint64_t tensor_size[kRank];
```
What's the difference between `cuuint64_t` and `uint64_t`?

hummingtree replied:
I have no idea ...

maddyscientist commented on include/targets/cuda/mma_tensor_op/shared_memory_pattern.cuh:
```diff
@@ -0,0 +1,354 @@
+  template <int tmp_ld, int smem_ld>
+  struct tmp2s_smem_t<float, float, true, tmp_ld, smem_ld, std::enable_if_t<(tmp_ld % 8 == 0) && (tmp_ld / 8) % 2 == 1, void>> {
+
+    static constexpr int get_wm() { return 8; }
```
Can you add documentation for what each of these functions returns?  No need to reproduce it for each specialization, could put it in the dummy on line 15.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on include/targets/cuda/mma_tensor_op/gmem_loader.cuh:
```diff
@@ -0,0 +1,802 @@
+        auto yy = p[(m_idx + 1) * ld + n_idx];
+
+        if constexpr (fixed) {
+          this_max = abs_max(scale_inv * xx.real(), this_max);
```
any benefit to deferring the multiply by `scale_inv` until after the max is found (since it is a constant)?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/53c8e31e71603e0bb70114ec4af6488a8eb5f828

maddyscientist commented on include/targets/cuda/mma_tensor_op/gmem_loader.cuh:
```diff
@@ -0,0 +1,802 @@
+          }
+        }
+
+        return 1.0f / block_rescale_factor;
```
I think this is another division that can be removed with precomputation of `1.0 / numeric_limits<half>::max()` on the host.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/53c8e31e71603e0bb70114ec4af6488a8eb5f828

maddyscientist commented on include/kernels/dslash_mdw_fused.cuh:
```diff
@@ -311,22 +311,22 @@ namespace quda {
         if (Arg::type == MdwfFusedDslashType::D4_D5INV_D5PRE) {
           if (arg.small_kappa) {
-            construct_matrix_a_d5<M_sm, false, Arg>(arg, sm_a); // dagger = false
+            construct_matrix_a_d5<M_sm, false, Arg>(arg, sm_a, arg.alpha, arg.beta); // dagger = false
```
What is this change to `construct_matrix_a_mg5inv` motivated by (e.g., the addition of explicit alpha and beta arguments)?

hummingtree replied:
In most cases `alpha = arg.alpha`, `beta = arg.beta`; but in a few others cases we need to use other values for `alpha` and `beta`. Previously this is done by modifying `arg.alpha` and `arg.beta`, which is a bad practice in the first place. Now that arg becomes const so we need to change the function interface to pass in `alpha` and `beta` explicitly.

maddyscientist commented on lib/check_params.h:
```diff
@@ -878,9 +878,11 @@ void printQudaMultigridParam(QudaMultigridParam *param) {
     P(setup_use_mma[i], QUDA_BOOLEAN_FALSE);
 #endif
     P(dslash_use_mma[i], QUDA_BOOLEAN_FALSE);
+    P(transfer_use_mma[i], QUDA_BOOLEAN_FALSE);
```
Meta comment: while eventually we will want the mma options to be default on, I think for now we should leave them opt in.  Switching to default on will come when I've got the entire coarse grid correction working in MMA order, avoiding the reordering overheads.

hummingtree replied:
For now only setup has the MMA option default to on.

maddyscientist commented on lib/restrictor_mma.in.cu:
```diff
@@ -0,0 +1,279 @@
+            RestrictMma<store_t, store_t, 4, fineColor, coarseColor, nVec>(out, in, v, fine_to_coarse, coarse_to_fine,
+                                                                           spin_map, parity);
+          } else if (in.Precision() == QUDA_HALF_PRECISION) {
+#if 0
```
Can you add some documentation here, describing why this is disabled at the moment?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on lib/restrictor_mma.in.cu:
```diff
@@ -0,0 +1,279 @@
+
+    using mma_t = typename mma::mg_mma_restrictor_t<out_t>::type;
+
+    static constexpr int spin_block_factor = spin_mapper<fineSpin, coarseSpin>::get_spin_block_factor();
```
some comments about the numerical values of `aggregate_size_block_max`, `block_atom_size` and `block_limit` would be helpful

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on lib/prolongator_mma.in.cu:
```diff
@@ -0,0 +1,236 @@
+    static constexpr int n_atom_size = mma_t::MMA_N;
+    static constexpr int m_atom_size = mma_t::MMA_M;
+    static constexpr int k_atom_size = mma_t::MMA_K;
+    static constexpr int block_atom_size = 32 / 8;
```
A comment about why the numerical value block_atom_size = 32 / 8 would be helpful

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist commented on lib/multigrid.in.hpp:
```diff
@@ -0,0 +1,51 @@
+  template <int... Values>
+  auto sort_values() {
+    std::array<int, sizeof...(Values)> arr = {Values...};
+    // std::sort is NOT constexpr until C++20
```
Will having `std::sort` as `constexpr` be helpful here?

hummingtree replied:
If `std::sort` becomes `constexpr`, we can add `constexpr` to the `auto sorted_nVecs = sort_values<@QUDA_MULTIGRID_MRHS_LIST@>();` below. I doubt it would give any speed up in practice though.

maddyscientist commented on lib/block_transpose.in.cu:
```diff
@@ -57,10 +60,19 @@ namespace quda
-        tp.set_max_shared_bytes = true;
-        resizeVector(tp.block.y * tp.grid.y); // We need a full threadblock
-        launch_device<BlockTransposeKernel>(tp, stream, arg);
+        if (from_to_non_rel) {
```
Should this be `if constexpr`?

hummingtree replied:
`from_to_non_rel` is not a template argument here, but the line right after this can add a `constexpr`.

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8

maddyscientist reviewed (changes requested):
Fantastic addition to QUDA @hummingtree and is a necessary step forward to future QUDA scalability.  I've left quite a few comments, but nothing too onerous to deal with I think.

hummingtree committed:
Add comments; remove dead code; address some of the review comments.

hummingtree committed:
Merge branch 'feature/prolongator-mma' of github.com:lattice/quda into feature/prolongator-mma

hummingtree committed:
Move the conditional macro to tma_helper.hpp

hummingtree committed:
Fix the plumbing for prolongator/restrictor-MMA for staggered.

hummingtree committed:
Use divide and conquer logic when choosing the instantiated nVec to use.

hummingtree committed:
Optimize the rescaling code:
- Replace division with multiplicaitons;
- Find maximum value before adding performing multiplication.

hummingtree commented:
> As far as I can tell it doesn't look like the MMA prolongator/restrictor is getting picked up with staggered on Hopper, at least using the default `cmake` parameters:
> 
> ```
> cmake -DQUDA_DIRAC_DEFAULT_OFF=ON \
>       -DQUDA_DIRAC_STAGGERED=ON \
>       -DCMAKE_BUILD_TYPE=DEVEL \
>       -DQUDA_BACKWARDS=ON \
>       -DCMAKE_INSTALL_PREFIX=/scratch/local/install \
>       -DQUDA_MULTIGRID=ON \
>       -DQUDA_MULTIGRID_NVEC_LIST="24,64,96" \
>       -DQUDA_MULTIGRID_MRHS_LIST="8,16,32" \
>       -DQUDA_DOWNLOAD_USQCD=ON \
>       -DQUDA_QIO=ON -DQUDA_QMP=ON \
>       -DQUDA_PRECISION=14 \
>       -DQUDA_RECONSTRUCT=4 \
>       -DQUDA_GPU_ARCH=sm_90 \
>       /scratch/local/quda
> ```
> 
> Commands:
> 
> ```
> export QUDA_RESOURCE_PATH=`pwd`/tunecache
> 
> mpirun -np 1 ./heatbath_test --dim 16 16 16 16 --save-gauge l16t16b7p0   --heatbath-beta 7.0 --heatbath-coldstart true --heatbath-num-steps 10 --heatbath-warmup-steps 1000
> 
> mpirun -np 1 ./staggered_invert_test \
>   --prec double --prec-sloppy single --prec-null half --prec-precondition half \
>   --mass 0.01 --recon 18 --recon-sloppy 18 --recon-precondition 18 \
>   --dim 16 16 16 16 --gridsize 1 1 1 1 --load-gauge l16t16b7p0 \
>   --dslash-type asqtad --compute-fat-long true --tadpole-coeff 0.905160183 --tol 1e-10 \
>   --verbosity verbose --solve-type direct --solution-type mat --inv-type gcr \
>   --inv-multigrid true --mg-levels 4 --mg-coarse-solve-type 0 direct --mg-staggered-coarsen-type kd-optimized \
>   --mg-block-size 0 1 1 1 1 --mg-nvec 0 3 \
>   --mg-block-size 1 4 4 4 4 --mg-nvec 1 64 --mg-nvec-batch 1 32 \
>   --mg-block-size 2 2 2 2 2 --mg-nvec 2 96 --mg-nvec-batch 2 32 \
>   --mg-setup-tol 1 1e-5 --mg-setup-tol 2 1e-5 --mg-setup-inv 1 cgnr --mg-setup-inv 2 cgnr \
>   --nsrc 32 --nsrc-tile 8 --niter 25 \
>   --mg-setup-use-mma 0 true --mg-setup-use-mma 1 true --mg-setup-use-mma 2 true --mg-setup-use-mma 3 true \
>   --mg-dslash-use-mma 0 true --mg-dslash-use-mma 1 true --mg-dslash-use-mma 2 true --mg-dslash-use-mma 3 true \
>   --mg-transfer-use-mma 0 true --mg-transfer-use-mma 1 true --mg-transfer-use-mma 2 true --mg-transfer-use-mma 3 true \
>   --mg-smoother 0 ca-gcr --mg-smoother-solve-type 0 direct  --mg-nu-pre 0 0 --mg-nu-post 0 4 \
>   --mg-smoother 1 ca-gcr --mg-smoother-solve-type 1 direct --mg-nu-pre 1 0 --mg-nu-post 1 4 \
>   --mg-smoother 2 ca-gcr --mg-smoother-solve-type 2 direct-pc  --mg-nu-pre 2 0 --mg-nu-post 2 4 \
>   --mg-coarse-solver 1 gcr --mg-coarse-solve-type 1 direct --mg-coarse-solver-tol 1 0.25 --mg-coarse-solver-maxiter 1 16 \
>   --mg-coarse-solver 2 gcr --mg-coarse-solve-type 2 direct-pc --mg-coarse-solver-tol 2 0.25 --mg-coarse-solver-maxiter 2 16 \
>   --mg-coarse-solver 3 ca-gcr --mg-coarse-solve-type 3 direct-pc --mg-coarse-solver-tol 3 0.25 --mg-coarse-solver-maxiter 3 16 \
>   --mg-verbosity 0 verbose --mg-verbosity 1 verbose --mg-verbosity 2 verbose --mg-verbosity 3 verbose
> ```
> 
> I can confirm the dslash and setup flags are being picked up correctly (but I was a bit surprised to see only `1xfp16` was getting used for the half-precision `CalculateYhat`, it felt a little low...), so I think it's uniquely the prolongator/restrictor that's having plumbing issues, at least for the default.

This has been fixed in https://github.com/lattice/quda/pull/1497/commits/85da7a329b41d2d8dd55330fed972f320a0db7fe.

hummingtree committed:
When using TF32 and BF16 check if the CC is less than 80.

hummingtree committed:
Apply the Arg::check_bounds to the other kernels.

hummingtree committed:
Further reduece the number of float division by using scale.

hummingtree committed:
Allow specify MMA types per precision.

hummingtree committed:
Break the MMA type into half/single ones.

hummingtree committed:
Print the nVec used only in debug verbosity.

maddyscientist commented on lib/dslash_coarse.in.cpp:
```diff
@@ -105,7 +105,9 @@ namespace quda
         int instantiated_nVec = instantiated_nVec_to_use(out.size());
         size_t size = out.size();
-        printfQuda("nVec/out.size() = %d/%lu\n", instantiated_nVec, size);
+        if (getVerbosity() >= QUDA_DEBUG_VERBOSE) {
```
Use `logQuda` here?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/f8235601e734cf2d84a731ad1fccbd9089372449

maddyscientist commented on lib/multigrid.in.hpp:
```diff
@@ -51,7 +51,9 @@ namespace quda {
     int instantiated_nVec = instantiated_nVec_to_use(out.size());
     size_t size = out.size();
-    printfQuda("divide_and_conquer nVec/out.size() = %d/%lu\n", instantiated_nVec, size);
+    if (getVerbosity() >= QUDA_DEBUG_VERBOSE) {
```
Use `logQuda` here?

hummingtree replied:
https://github.com/lattice/quda/pull/1497/commits/f8235601e734cf2d84a731ad1fccbd9089372449

maddyscientist reviewed (approved):
Thanks for the requested fixes done on this @hummingtree.  Aside from a trivial comment I just made (`logQuda`) this is good to go as far as I am concerned.

weinbe2 commented:
Good news: this passes a visual review! Bad news: I hit an issue that's only present with `--mg-dslash-use-mma` enabled, single precision, Nc = 96... **and** it goes away if I disable auto-tuning, so I've attached the tunecache as well... This is on Hopper SXM 80GB. I'm not sure if I can trigger it with a single GPU build since it's tuning specific...

`cmake` command:
```
cmake -DCMAKE_BUILD_TYPE=RELEASE -DQUDA_DIRAC_DEFAULT_OFF=ON -DQUDA_DIRAC_STAGGERED=ON \
  -DQUDA_GPU_ARCH=sm_90 -DQUDA_DOWNLOAD_USQCD=ON -DQUDA_QIO=ON -DQUDA_QMP=ON \
  -DQUDA_PRECISION=4 -DQUDA_RECONSTRUCT=4 \
  -DQUDA_MULTIGRID=ON -DQUDA_MULTIGRID_NVEC_LIST="24,64,96" -DQUDA_MULTIGRID_MRHS_LIST="8,16,32" \
  /scratch/local/quda
```

Command---with the tunecache I have, it only triggers with single precision, `--mg-dslash-use-mma 3 true`, 4 level solve... and the issue only hits on the coarsest level; printout of error below.

```
PREC="single"

mpirun -np 1 ./staggered_invert_test \
  --prec single --prec-sloppy single --prec-null $PREC --prec-precondition $PREC \
  --mass 0.2 --recon 18 --recon-sloppy 18 --recon-precondition 18 \
  --dim 16 16 16 16 --gridsize 1 1 1 1 \
  --dslash-type staggered --compute-fat-long false --tadpole-coeff 0.905160183 --tol 1e-10 \
  --verbosity verbose --solve-type direct --solution-type mat --inv-type gcr \
  --inv-multigrid true --mg-levels 4 --mg-coarse-solve-type 0 direct --mg-staggered-coarsen-type kd-optimized \
  --mg-block-size 0 1 1 1 1 --mg-nvec 0 3 \
  --mg-block-size 1 4 4 4 4 --mg-nvec 1 64 --mg-nvec-batch 1 32 \
  --mg-block-size 2 2 2 2 2 --mg-nvec 2 96 --mg-nvec-batch 2 32 \
  --mg-setup-tol 1 1e-5 --mg-setup-tol 2 1e-5 --mg-setup-inv 1 cgnr --mg-setup-inv 2 cgnr \
  --nsrc 32 --nsrc-tile 16 --niter 24 \
  --mg-setup-use-mma 0 true --mg-setup-use-mma 1 true --mg-setup-use-mma 2 true --mg-setup-use-mma 3 true \
  --mg-dslash-use-mma 0 true --mg-dslash-use-mma 1 true --mg-dslash-use-mma 2 true --mg-dslash-use-mma 3 true \
  --mg-transfer-use-mma 0 false --mg-transfer-use-mma 1 false --mg-transfer-use-mma 2 false --mg-transfer-use-mma 3 false \
  --mg-smoother 0 ca-gcr --mg-smoother-solve-type 0 direct  --mg-nu-pre 0 0 --mg-nu-post 0 4 \
  --mg-smoother 1 ca-gcr --mg-smoother-solve-type 1 direct --mg-nu-pre 1 0 --mg-nu-post 1 4 \
  --mg-smoother 2 ca-gcr --mg-smoother-solve-type 2 direct-pc  --mg-nu-pre 2 0 --mg-nu-post 2 4 \
  --mg-coarse-solver 1 gcr --mg-coarse-solve-type 1 direct --mg-coarse-solver-tol 1 0.25 --mg-coarse-solver-maxiter 1 16 \
  --mg-coarse-solver 2 gcr --mg-coarse-solve-type 2 direct-pc --mg-coarse-solver-tol 2 0.25 --mg-coarse-solver-maxiter 2 16 \
  --mg-coarse-solver 3 ca-gcr --mg-coarse-solve-type 3 direct-pc --mg-coarse-solver-tol 3 0.25 --mg-coarse-solver-maxiter 3 16 \
  --mg-verbosity 0 verbose --mg-verbosity 1 verbose --mg-verbosity 2 verbose --mg-verbosity 3 verbose
```

Here's the error, it's CA-GCR on the coarsest level very quickly, after the first norm check after a dslash; it also breaks with any other solver, so it seems like it's the coarsest dslash itself. It's unique to a batched solve, seemingly independent of `--nsrc-tile`? So maybe there's some weird corner of parameter space. I can only trigger it with Nc = 96 on the coarsest level, if it's on the intermediate level (replace `--mg-nvec 1 64` with `96`) things are fine...

```
[...]
MG level 2 (GPU): GCR:     0 iterations, n = 8, <r,r> = 4.825448e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 9, <r,r> = 4.814075e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 10, <r,r> = 4.881138e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 11, <r,r> = 4.742637e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 12, <r,r> = 4.829302e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 13, <r,r> = 4.812252e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 14, <r,r> = 4.755250e+04, |r|/|b| = 1.000000e+00
MG level 2 (GPU): GCR:     0 iterations, n = 15, <r,r> = 4.762029e+04, |r|/|b| = 1.000000e+00
MG level 3 (GPU): CA-GCR:     0 iterations, n = 0, <r,r> =       nan, |r|/|b| =       nan
MG level 3 (GPU): ERROR: Solver appears to have diverged for n = 0 (rank 0, host ipp2-0709.nvidia.com, solver.cpp:479 in void quda::Solver::PrintStats(const char*, int, quda::cvector<double>&, quda::cvector<double>&, quda::cvector<double>&)())
MG level 3 (GPU):        last kernel called was (name=cudaMemsetAsync,volume=bytes=12288,aux=zero,color_spinor_field.cpp,409)
MG level 3 (GPU): Saving 1207 sets of cached parameters to /scratch/local/build/tests/tunecache/tunecache_error.tsv
```

Reference tunecache: [tunecache_fail.tar.gz](https://github.com/user-attachments/files/18430765/tunecache_fail.tar.gz)

Commit id: 49c0a583acf99faf106d96041762e9b6d9f7b60a

weinbe2 commented:
Infinitely cleaner command... thanks @hummingtree 

```
            for PREC in half single
            do
                mpirun -n 1 ./multigrid_benchmark_test --test 0 --dim 2 2 2 2 --niter 10 --nsrc 8 --prec-sloppy ${PREC} --mg-nvec 0 96 --mg-dslash-use-mma 0 true
            done
```

hummingtree committed:
Remove the divide and conquer code for larger than limit TMA box sizes as they do not work; Add the logic to make sure the box sizes are not larger than the limit.

hummingtree committed:
Change the default setup MMA types to 3xfp16.

hummingtree committed:
Use logQuda.

hummingtree commented:
> Good news: this passes a visual review! Bad news: I hit an issue that's only present with `--mg-dslash-use-mma` enabled, single precision, Nc = 96... **and** it goes away if I disable auto-tuning, so I've attached the tunecache as well... This is on Hopper SXM 80GB. I'm not sure if I can trigger it with a single GPU build since it's tuning specific...
> 
> `cmake` command:
> 
> ```
> cmake -DCMAKE_BUILD_TYPE=RELEASE -DQUDA_DIRAC_DEFAULT_OFF=ON -DQUDA_DIRAC_STAGGERED=ON \
>   -DQUDA_GPU_ARCH=sm_90 -DQUDA_DOWNLOAD_USQCD=ON -DQUDA_QIO=ON -DQUDA_QMP=ON \
>   -DQUDA_PRECISION=4 -DQUDA_RECONSTRUCT=4 \
>   -DQUDA_MULTIGRID=ON -DQUDA_MULTIGRID_NVEC_LIST="24,64,96" -DQUDA_MULTIGRID_MRHS_LIST="8,16,32" \
>   /scratch/local/quda
> ```
> 
> Command---with the tunecache I have, it only triggers with single precision, `--mg-dslash-use-mma 3 true`, 4 level solve... and the issue only hits on the coarsest level; printout of error below.
> 
> ```
> PREC="single"
> 
> mpirun -np 1 ./staggered_invert_test \
>   --prec single --prec-sloppy single --prec-null $PREC --prec-precondition $PREC \
>   --mass 0.2 --recon 18 --recon-sloppy 18 --recon-precondition 18 \
>   --dim 16 16 16 16 --gridsize 1 1 1 1 \
>   --dslash-type staggered --compute-fat-long false --tadpole-coeff 0.905160183 --tol 1e-10 \
>   --verbosity verbose --solve-type direct --solution-type mat --inv-type gcr \
>   --inv-multigrid true --mg-levels 4 --mg-coarse-solve-type 0 direct --mg-staggered-coarsen-type kd-optimized \
>   --mg-block-size 0 1 1 1 1 --mg-nvec 0 3 \
>   --mg-block-size 1 4 4 4 4 --mg-nvec 1 64 --mg-nvec-batch 1 32 \
>   --mg-block-size 2 2 2 2 2 --mg-nvec 2 96 --mg-nvec-batch 2 32 \
>   --mg-setup-tol 1 1e-5 --mg-setup-tol 2 1e-5 --mg-setup-inv 1 cgnr --mg-setup-inv 2 cgnr \
>   --nsrc 32 --nsrc-tile 16 --niter 24 \
>   --mg-setup-use-mma 0 true --mg-setup-use-mma 1 true --mg-setup-use-mma 2 true --mg-setup-use-mma 3 true \
>   --mg-dslash-use-mma 0 true --mg-dslash-use-mma 1 true --mg-dslash-use-mma 2 true --mg-dslash-use-mma 3 true \
>   --mg-transfer-use-mma 0 false --mg-transfer-use-mma 1 false --mg-transfer-use-mma 2 false --mg-transfer-use-mma 3 false \
>   --mg-smoother 0 ca-gcr --mg-smoother-solve-type 0 direct  --mg-nu-pre 0 0 --mg-nu-post 0 4 \
>   --mg-smoother 1 ca-gcr --mg-smoother-solve-type 1 direct --mg-nu-pre 1 0 --mg-nu-post 1 4 \
>   --mg-smoother 2 ca-gcr --mg-smoother-solve-type 2 direct-pc  --mg-nu-pre 2 0 --mg-nu-post 2 4 \
>   --mg-coarse-solver 1 gcr --mg-coarse-solve-type 1 direct --mg-coarse-solver-tol 1 0.25 --mg-coarse-solver-maxiter 1 16 \
>   --mg-coarse-solver 2 gcr --mg-coarse-solve-type 2 direct-pc --mg-coarse-solver-tol 2 0.25 --mg-coarse-solver-maxiter 2 16 \
>   --mg-coarse-solver 3 ca-gcr --mg-coarse-solve-type 3 direct-pc --mg-coarse-solver-tol 3 0.25 --mg-coarse-solver-maxiter 3 16 \
>   --mg-verbosity 0 verbose --mg-verbosity 1 verbose --mg-verbosity 2 verbose --mg-verbosity 3 verbose
> ```
> 
> Here's the error, it's CA-GCR on the coarsest level very quickly, after the first norm check after a dslash; it also breaks with any other solver, so it seems like it's the coarsest dslash itself. It's unique to a batched solve, seemingly independent of `--nsrc-tile`? So maybe there's some weird corner of parameter space. I can only trigger it with Nc = 96 on the coarsest level, if it's on the intermediate level (replace `--mg-nvec 1 64` with `96`) things are fine...
> 
> ```
> [...]
> MG level 2 (GPU): GCR:     0 iterations, n = 8, <r,r> = 4.825448e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 9, <r,r> = 4.814075e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 10, <r,r> = 4.881138e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 11, <r,r> = 4.742637e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 12, <r,r> = 4.829302e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 13, <r,r> = 4.812252e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 14, <r,r> = 4.755250e+04, |r|/|b| = 1.000000e+00
> MG level 2 (GPU): GCR:     0 iterations, n = 15, <r,r> = 4.762029e+04, |r|/|b| = 1.000000e+00
> MG level 3 (GPU): CA-GCR:     0 iterations, n = 0, <r,r> =       nan, |r|/|b| =       nan
> MG level 3 (GPU): ERROR: Solver appears to have diverged for n = 0 (rank 0, host ipp2-0709.nvidia.com, solver.cpp:479 in void quda::Solver::PrintStats(const char*, int, quda::cvector<double>&, quda::cvector<double>&, quda::cvector<double>&)())
> MG level 3 (GPU):        last kernel called was (name=cudaMemsetAsync,volume=bytes=12288,aux=zero,color_spinor_field.cpp,409)
> MG level 3 (GPU): Saving 1207 sets of cached parameters to /scratch/local/build/tests/tunecache/tunecache_error.tsv
> ```
> 
> Reference tunecache: [tunecache_fail.tar.gz](https://github.com/user-attachments/files/18430765/tunecache_fail.tar.gz)
> 
> Commit id: [49c0a58](https://github.com/lattice/quda/commit/49c0a583acf99faf106d96041762e9b6d9f7b60a)

Thanks Evan for the tests! This should have been fixed in https://github.com/lattice/quda/pull/1497/commits/e8ca86924e6ed917717f4c062e3a2573d6ff11a5.

hummingtree committed:
Change from - uses: actions/checkout@v3 to - uses: actions/checkout@v4.

mathiaswagner committed:
Update cuda_githubactions_build.yml

attempt to fix build

weinbe2 reviewed (approved):
With the recent bugfixes this is a go---the recent issue I filed is orthogonal to this work. Awesome work @hummingtree !

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/prolongator-mma

hummingtree committed:
Not generating the Nc=6 files for MMA transfers.

hummingtree committed:
Disable compiling for fineColor=6 and coarseColor=6 for the transfer kernels.

hummingtree committed:
Apply clang-format.

hummingtree committed:
Resolve the cmake/clang-format conflicts, hopefully.

weinbe2 commented:
cscs-ci run

hummingtree committed:
Whitelist instead of blacklisting the fineColor and coarseColor for the transfers.

weinbe2 committed:
Added apt-get install clang-14 to github actions
